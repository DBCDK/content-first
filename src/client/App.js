import React, {Component} from 'react';
import {connect} from 'react-redux';
import {ToastContainer} from 'react-toastify';
import 'react-toastify/dist/ReactToastify.css';
import {isMobile} from 'react-device-detect';

import Head from './components/base/Head';
import Modal from './components/modals/Modal.container';
import FrontPage from './components/frontpage/FrontPage.container';
import FilterPage from './components/filter/FilterPage/FilterPage.container';
import WorkPage from './components/work/WorkPage/WorkPage.container';
import EditProfilePage from './components/profile/EditProfilePage';
import TopBar from './components/top/TopBar.component';
import {ON_USER_DETAILS_REQUEST} from './redux/user.reducer';
import ListPage from './components/list/ListPage';
import ShortList from './components/list/shortlist/ShortList.container';
import Spinner from './components/general/Spinner/Spinner.component';
import Styleguide from './components/Styleguide/Styleguide.component';
import FeedbackButton from './components/general/FeedbackButton/FeedbackButton.component';
import Footer from './components/general/Footer/Footer.component';
import Article from './components/article/Article.component';
import Animate from './components/base/Animate';
import CookieWarning from './components/general/CookieWarning/CookieWarning';
import BeltEditor from './components/belteditor/BeltEditor.component';
import BeltForm from './components/belteditor/BeltForm.component';
import PrintLayout from './components/list/printLayout/PrintLayout';

// kiosk
import KioskSetup from './components/kiosk/Setup/KioskSetup.component';
import Navigation from './components/kiosk/Navigation/Navigation.component';
import Kiosk from './components/base/Kiosk/Kiosk';
import KioskWorkPage from './components/kiosk/WorkPage/KioskWorkPage.component';
import Logo from './components/kiosk/Logo/Logo';
import Timeout from './components/kiosk/Timeout/Timeout';

import {OPEN_MODAL} from './redux/modal.reducer';

import './style/App.css';
import './style/index.css';
import {ADMIN_ROLE, EDITOR_ROLE} from './components/roles/Role.component';
import Role from './components/roles/Role.component';

class App extends Component {
  constructor(props) {
    super(props);
    this.state = {didShowCookieModal: false};
  }

  componentDidMount() {
    this.props.userDetailsRequest();
    this.screenHeight = window.innerHeight;
    this.offset = this.screenHeight / 6;

    window.addEventListener('resize', this.handleResize);
  }

  componentWillUnmount() {
    window.removeEventListener('resize', this.handleResize);
  }

  handleResize = () => {
    const softKeyboard =
      window.innerHeight < this.screenHeight - this.offset ? true : false;
    if (softKeyboard !== this.state.softKeyboard) {
      this.setState({softKeyboard});
    }
  };

  render() {
    if (!navigator.cookieEnabled && !this.state.didShowCookieModal) {
      this.props.cookieModal();
      this.setState({didShowCookieModal: true});
    }

    const isTouchClass = isMobile ? 'isTouch' : '';
    const isPremiumClass = this.props.isPremium ? 'premium' : '';
    const isKioskClass = this.props.isKiosk ? 'kioskmode' : '';
    const softKeyboardClass =
      this.props.isKiosk && this.state.softKeyboard ? 'keyboard' : '';

    const path = this.props.routerState.path;
    const pathSplit = path.split('/');

    const backgroundColor =
      pathSplit && pathSplit[1] === 'styleguide'
        ? 'var(--lys-graa)'
        : 'var(--white)';

    let currentPage = null;
    let topbar = true;
    let footer = true;
    let feedBack = true;
    if (pathSplit[1] === '') {
      currentPage = <FrontPage />;
    } else if (pathSplit[1] === 'værk') {
      if (this.props.isKiosk) {
        currentPage = <KioskWorkPage pid={pathSplit[2]} />;
      } else {
        currentPage = <WorkPage pid={pathSplit[2]} />;
      }
    } else if (pathSplit[1] === 'profil') {
      if (pathSplit[2] === 'rediger') {
        currentPage = <EditProfilePage />;
      }
    } else if (pathSplit[1] === 'lister') {
      if (pathSplit[2]) {
        currentPage = <ListPage id={pathSplit[2]} />;
      }
    } else if (pathSplit[1] === 'huskeliste') {
      currentPage = <ShortList />;
    } else if (pathSplit[1] === 'find') {
      currentPage = <FilterPage />;
    } else if (pathSplit[1] === 'replay') {
      currentPage = (
        <Spinner
          className="d-block mx-auto"
          style={{marginTop: 200}}
          size="50px"
        />
      );
    } else if (pathSplit[1] === 'redaktionen') {
      if (pathSplit[2] === 'opret') {
        currentPage = (
          <Role
            requiredRoles={[ADMIN_ROLE, EDITOR_ROLE]}
            onAccessDenied={
              <Article path={`/${pathSplit[1]}/${pathSplit[2]}`} />
            }
          >
            <BeltForm mode="create" />
          </Role>
        );
      } else if (pathSplit[2] === 'rediger') {
        currentPage = (
          <Role
            requiredRoles={[ADMIN_ROLE, EDITOR_ROLE]}
            onAccessDenied={
              <Article path={`/${pathSplit[1]}/${pathSplit[2]}`} />
            }
          >
            <BeltForm
              mode="edit"
              {...this.beltFormParams(this.props.routerState.params)}
            />
          </Role>
        );
      } else {
        currentPage = (
          <Role
            requiredRoles={[ADMIN_ROLE, EDITOR_ROLE]}
            onAccessDenied={<Article path={`/${pathSplit[1]}`} />}
          >
            <BeltEditor query={{type: 'belt', owner: EDITOR_ROLE}} />
          </Role>
        );
      }
    } else if (pathSplit[1] === 'styleguide') {
      currentPage = <Styleguide />;
      footer = false;
    } else if (pathSplit[1] === 'print' && pathSplit[2]) {
      currentPage = <PrintLayout id={pathSplit[2]} />;
      topbar = false;
      footer = false;
      feedBack = false;
    } else if (pathSplit[1] === 'kiosk') {
      currentPage = <KioskSetup />;
    }

    if (!currentPage) {
      currentPage = <Article path={`/${pathSplit[1]}`} />;
    }

    return (
      <div
        id="App"
        className={`App ${isKioskClass} ${softKeyboardClass} ${isPremiumClass} ${isTouchClass}`}
        style={{backgroundColor}}
      >
        <Head />

        {this.props.isKiosk && <Timeout />}

        {this.props.isKiosk && <Logo />}

        <Kiosk
          render={({kiosk}) => {
            if (kiosk.enabled) {
              return <Navigation />;
            }
            return (
              topbar && (
                <div>
                  <TopBar
                    dispatch={this.props.dispatch}
                    user={this.props.user}
                  />
                  <div className="App__TopbarPlaceholder" />
                </div>
              )
            );
          }}
        />
        <div id="scrollableArea">{currentPage}</div>
        <Modal />
        {!this.props.isKiosk && <CookieWarning />}
        <Animate />

        <ToastContainer
          position="top-right"
          autoClose={5000}
          hideProgressBar={false}
          newestOnTop
          closeOnClick
          rtl={false}
          pauseOnVisibilityChange={false}
          draggable
          pauseOnHover={false}
          delay={5000}
        />

        {feedBack && !this.props.isKiosk && <FeedbackButton />}

        <Kiosk
          render={({kiosk}) => {
            if (kiosk.enabled) {
              return <div className="navigation--placeholder" />;
            }
            return footer && <Footer />;
          }}
        />
      </div>
    );
  }

  beltFormParams = params => ({
    title: params.title
      ? Array.isArray(params.title[0])
        ? params.title[0].join()
        : params.title[0]
      : '',
    description: params.description
      ? Array.isArray(params.description[0])
        ? params.description[0].join()
        : params.description[0]
      : '',
    enabled: params.enabled ? params.enabled[0] : false,
    tags: params.tags ? params.tags[0] : [],
    createdBy: params.createdBy ? params.createdBy[0] : '',
    created: params.created ? params.created[0] : '',
    id: params.id ? params.id[0] : '',
    index: params.index ? params.index[0] : ''
  });
}

const mapStateToProps = state => {
  return {
    routerState: state.routerReducer,
    isKiosk: state.kiosk.enabled,
    user: state.userReducer,
    isPremium: state.userReducer.isPremium
  };
};

export const mapDispatchToProps = dispatch => ({
  userDetailsRequest: () => dispatch({type: ON_USER_DETAILS_REQUEST}),
  cookieModal: () => {
    dispatch({
      type: OPEN_MODAL,
      modal: 'confirm',
      context: {
        title: 'COOKIES ER SLÅET FRA',
        reason:
          'Din browser tillader ikke cookies, og det betyder, at dele af Læsekompas.dk ikke vil virke. ' +
          'Vi anbefaler, at du ændrer indstillingen i din browser og tillader cookies, så du kan få den fulde oplevelse og de bedste læseforslag her på siden.',
        confirmText: 'Ok',
        hideCancel: true,
        onConfirm: () => {
          dispatch({
            type: 'CLOSE_MODAL',
            modal: 'confirm'
          });
        },
        onCancel: () => {
          dispatch({
            type: 'CLOSE_MODAL',
            modal: 'confirm'
          });
        }
      }
    });
  }
});
export default connect(
  // Map redux state to props
  mapStateToProps,
  mapDispatchToProps
)(App);
