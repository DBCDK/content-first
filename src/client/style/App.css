.App {
  min-height: 100vh;
  position: relative;
  background: white;
}

.App__TopbarPlaceholder {
  height: 80px;
}

.App.kioskmode .Toolbar button.AddToListButton,
.App.kioskmode .Toolbar button.Ebog,
.App.kioskmode .Toolbar button.Lydbog,
.App.kioskmode .Toolbar button.orderAllBtn,
.App.kioskmode .Toolbar button.printListBtn,
.App.kioskmode .Toolbar p.loanTitle {
  /* display: none; */
}

/* Kioskmode enabled */
/* hides all blue outline borders on button/input focus */
.App.kioskmode *:focus {
  outline: none;
}

@media (min-width: 600px) {
  .App {
    min-width: 100%;
  }

  .container {
    min-width: calc(100% - 100px);
  }
}

@media only screen and (max-width: 575px) {
  .App__TopbarPlaceholder {
    height: 60px;
  }
}

@media (min-width: 2000px) {
  .App {
    min-width: 2000px;
  }
}
