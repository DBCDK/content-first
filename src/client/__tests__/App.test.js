import React from 'react';
import App from '../App';
import {Provider} from 'react-redux';
import createStore from '../redux/Store';
import renderer from 'react-test-renderer';

// jest.mock('../../data/exportTaxonomy.json', () => {
//   return require('./__mocks__/mockedTaxonomy.json');
// });

jest.mock('../components/base/Belt/Slider.component', () => 'slider');
jest.mock('../components/base/Belt/CreatorBelt.component', () => 'CreatorBelt');
jest.mock('../components/base/Belt/TagsBelt.component', () => 'TagsBelt');
jest.mock(
  '../components/base/Belt/RecentListsBelt.container',
  () => 'RecentListsBelt'
);
jest.mock('../components/spots/Spots.container.js', () => 'SpotsContainer');
jest.mock('../components/hero/Hero.component.js', () => 'Hero');
jest.mock('../components/general/Footer/Footer.component.js', () => 'Footer');

it('renders without crashing', () => {
  window.$ = () => ({tooltip: () => {}}); // mock jquery tooltip
  const tree = renderer
    .create(
      <Provider store={createStore()}>
        <App />
      </Provider>
    )
    .toJSON();
  expect(tree).toMatchSnapshot();
});
