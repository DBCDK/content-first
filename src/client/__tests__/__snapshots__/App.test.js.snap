// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`renders without crashing 1`] = `
<div
  className="App container"
>
  <div
    className="row topbar"
  >
    <div
      className="col-xs-6 text-left header"
      onClick={[Function]}
    >
      <div>
        <img
          alt="logo"
          src="logo.svg"
        />
      </div>
      <div>
        <h1>
          Læsekompasset
        </h1>
      </div>
    </div>
    <div
      className="col-xs-6 text-right login"
    >
      <div
        className="short-list"
      >
        <div
          className="short-list--btn"
          onClick={[Function]}
        >
          <img
            alt=""
            src="Huskeliste.svg"
            title="Huskeliste"
          />
        </div>
        <div
          className="short-list--content text-left slide-out"
        >
          <img
            alt="luk"
            className="short-list--close-btn"
            onClick={[Function]}
            src="Kryds.svg"
          />
          <div
            className="short-list--header"
          >
            HUSKELISTE
          </div>
          <div
            className="short-list--empty-text text-center"
          >
            Din huskeliste er tom
          </div>
          <div
            className="short-list--more-btn text-center"
          >
            <span
              className="btn btn-default text-center"
              onClick={[Function]}
            >
              SE HELE LISTEN
            </span>
          </div>
        </div>
      </div>
      <span
        onClick={[Function]}
      >
        Log ind
      </span>
    </div>
  </div>
  <div
    style={
      Object {
        "height": "50px",
      }
    }
  />
  <div
    className="frontpage"
  >
    <div
      className="row frontpage-image"
    />
    <div
      className="row frontpage-image-credits text-right"
    >
      <a
        href="https://www.flickr.com/people/fatseth/"
      >
        Foto ©G Morel
      </a>
    </div>
    <div
      className="belts col-xs-11 col-centered"
    >
      <div
        className="row belt text-left"
      >
        <div
          className="col-xs-12 header"
        >
          <span
            className="belt-title"
            data-html="true"
            data-original-title="<span>Medium længde</span> <span>Er på mange biblioteker</span> <span>Medrivende</span> <span>Udlånes meget</span>"
            data-toggle="tooltip"
            onClick={[Function]}
          >
            En god bog
          </span>
        </div>
        <div
          className="col-xs-12 belt-links"
        >
          <span>
            Se også:
          </span>
          <span
            className="belt-link"
            data-html="true"
            data-original-title="<span>Spændende</span> <span>Høj fart</span> <span>Plot</span>"
            data-toggle="tooltip"
            onClick={[Function]}
          >
            En spændende bog
          </span>
          <span
            className="belt-link"
            data-html="true"
            data-original-title="<span>Alternativ</span> <span>Innovativ</span>"
            data-toggle="tooltip"
            onClick={[Function]}
          >
            En anderledes bog
          </span>
        </div>
        <div
          className="row mb4"
        >
          <div
            className="col-xs-12"
          >
            <slider />
          </div>
        </div>
      </div>
      <div
        className="row belt text-left"
      >
        <div
          className="col-xs-12 header"
        >
          <span
            className="belt-title"
            data-html="true"
            data-original-title="<span>Ligefrem</span> <span>I bevægelse</span>"
            data-toggle="tooltip"
            onClick={[Function]}
          >
            En let læst bog
          </span>
        </div>
        <div
          className="col-xs-12 belt-links"
        >
          <span>
            Se også:
          </span>
          <span
            className="belt-link"
            data-html="true"
            data-original-title="<span>Filosofisk</span> <span>Indsigtsfuld</span>"
            data-toggle="tooltip"
            onClick={[Function]}
          >
            En bog der gør dig klogere
          </span>
          <span
            className="belt-link"
            data-html="true"
            data-original-title="<span>Meget åben for fortolkning</span> <span>Krævende</span>"
            data-toggle="tooltip"
            onClick={[Function]}
          >
            En udfordrende bog
          </span>
        </div>
        <div
          className="row mb4"
        >
          <div
            className="col-xs-12"
          >
            <slider />
          </div>
        </div>
      </div>
      <div
        className="row belt text-left"
      >
        <div
          className="col-xs-12 header"
        >
          <span
            className="belt-title"
            data-html="true"
            data-original-title="<span>Bibliotekaren anbefaler</span>"
            data-toggle="tooltip"
            onClick={[Function]}
          >
            Bibliotekarens ugentlige anbefalinger
          </span>
        </div>
        <div
          className="col-xs-12 belt-links"
        />
        <div
          className="row mb4"
        >
          <div
            className="col-xs-12"
          >
            <slider />
          </div>
        </div>
      </div>
      <div
        className="row belt text-left"
      >
        <div
          className="col-xs-12 header"
        >
          <span
            className="belt-title"
            data-html="true"
            data-original-title="<span>Ingen filtre</span>"
            data-toggle="tooltip"
            onClick={[Function]}
          >
            Passer med min smag
          </span>
        </div>
        <div
          className="col-xs-12 belt-links"
        />
        <div
          style={
            Object {
              "paddingLeft": "15px",
              "paddingRight": "15px",
            }
          }
        >
          <div
            className="create-profile col-xs-12 text-center"
          >
            <button
              className="btn btn-success"
            >
              Opret profil
            </button>
          </div>
        </div>
      </div>
      <div
        className="row belt text-left"
      >
        <div
          className="col-xs-12 header"
        >
          <span
            className="belt-title"
          >
            Seneste lister fra brugerne
          </span>
        </div>
        <div
          className="row mb4"
        >
          <div
            className="col-xs-12"
          >
            <slider />
          </div>
        </div>
      </div>
    </div>
  </div>
  <span />
</div>
`;
